File: 1481-Least_Number_of_Unique_Integers_after_K_Removals/Least_Number_of_Unique_Integers_after_K_Removals.go
package leastnumberofuniqueintegersafterkremovals

func findLeastNumOfUniqueInts(arr []int, k int) int {
	freqsMap := make(map[int]int)

	for _, key := range arr {
		val := freqsMap[key]
		freqsMap[key] = val + 1
	}

	freqs := make([]int, len(arr)+1)

	for _, freq := range freqsMap {
		freqs[freq]++
	}

	remainingUnique := len(freqsMap)
	for i := range freqs {
		for freqs[i] > 0 {
			if k >= i {
				k -= i
				freqs[i]--
				remainingUnique--
			} else {
				return remainingUnique
			}
		}
	}

	return remainingUnique
}

File: 1481-LeastNumberOfUniqueIntegersAfterKRemovals/LeastNumberOfUniqueIntegersAfterKRemovals.java
import java.util.*;

class LeastNumberOfUniqueIntegersAfterKRemovals {
    
    class Solution {
        public int findLeastNumOfUniqueInts(int[] arr, int k) {
            HashMap<Integer, Integer> hm = new HashMap<>();
            
            for(int i = 0; i < arr.length; i++){
                hm.put(arr[i], hm.getOrDefault(arr[i], 0) + 1);
            }
            
            
            int[] frequencies = new int[arr.length + 1];
            for(Integer freq: hm.values()){
                frequencies[freq]++;
            }
            
            int ans = hm.size();
            for(int i = 1; i < frequencies.length; i++){
                while(frequencies[i] > 0) {
                    if(k >= i) {
                        k -= i;
                        frequencies[i]--;
                        ans --;
                    }
                    else {
                        return ans;
                    }
                }
            }
            
            return ans;
            
        }   
    }
}
